File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/qr/QRDetectorStep.java
package hu.kdea.szavazas.ballotprocessor.qr;

import boofcv.struct.image.GrayU8;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;

public class QRDetectorStep {
    private final ZXingQRProcessor qrProcessor;

    @Inject
    public QRDetectorStep(ZXingQRProcessor qrProcessor) {
        this.qrProcessor = qrProcessor;
    }

    public QrData detect(GrayU8 scaledGray) throws InterruptedException {
        final QrResult[] detected = new QrResult[1];
        CountDownLatch latch = new CountDownLatch(1);
        qrProcessor.apply(scaledGray, result -> {
            detected[0] = result;
            latch.countDown();
        });
        if (latch.await(5, TimeUnit.SECONDS) && detected[0] != null) {
            return new QrData(
                detected[0].raw(),
                detected[0].numSupport(),
                detected[0].numCandidates(),
                detected[0].boundingBox()
            );
        }
        return null;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/QRDetectorStep.kt
// QRDetectorStep.kt
package hu.kdea.szavazas

import boofcv.struct.image.GrayU8
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit

class QRDetectorStep(private val qrProcessor: IQRProcessor) {
    fun detect(scaledGray: GrayU8): QrData? {
        var detected: QrResult? = null
        val latch = CountDownLatch(1)
        qrProcessor.detect(scaledGray) { r ->
            detected = r
            latch.countDown()
        }
        return if (latch.await(5, TimeUnit.SECONDS) && detected != null) {
            QrData(detected!!.raw, detected!!.numSupport, detected!!.numCandidates, detected!!.boundingBox)
        } else null
    }
}